File: src/editor/gui/generic/TristateCheckBox.java
package editor.gui.generic;

import java.awt.Color;
import java.awt.Component;
import java.awt.Graphics;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

import javax.swing.Icon;
import javax.swing.JCheckBox;
import javax.swing.UIManager;

/**
 * This class represents a check box with three states: selected, unselected,
 * and intermediate. A state of intermediate should be used to indicate when the
 * box represents an aspect of several items and not all of those items have the
 * that aspect.
 * 
 * @author Alec Roelke
 */
@SuppressWarnings("serial")
public class TristateCheckBox extends JCheckBox
{
    /**
     * The three states a TristateCheckBox can have.
     */
    public static enum State { SELECTED, INDETERMINATE, UNSELECTED }

    /**
     * Icon serving as the base for a TristateCheckBox.  A square is drawn
     * over it if the box is in indeterminate state.
     */
    private static final Icon BASE_ICON = UIManager.getIcon("CheckBox.icon");

    /**
     * Amount of the image the square should fill.
     */
    private static final double ICON_FILL = 2.0/3.0;

    /**
     * When the check box is clicked, toggle its state.  If the state is
     * indeterminate, change it to being unselected.
     */
    private class ClickListener implements ActionListener
    {
        @Override
        public void actionPerformed(ActionEvent e)
        {
            switch (state)
            {
            case SELECTED: case INDETERMINATE:
                setState(State.UNSELECTED);
                break;
            case UNSELECTED:
                state = State.SELECTED;
                setState(State.SELECTED);
                break;
            }
        }
    }

    /**
     * Icon for a TristateCheckBox.  Draws a normal check box icon, but
     * overlays it with a square if the box is in indeterminate state.
     */
    private class IndeterminateIcon implements Icon
    {
        /**
         * Color of the square to draw.
         */
        private Color boxColor;

        /**
         * Create a new IndeterminateIcon that draws a square of the specified
         * color.
         * 
         * @param color color of the square
         */
        public IndeterminateIcon(Color color)
        {
            boxColor = color;
        }

        @Override
        public int getIconHeight()
        {
            return BASE_ICON.getIconHeight();
        }

        @Override
        public int getIconWidth()
        {
            return BASE_ICON.getIconWidth();
        }

        @Override
        public void paintIcon(Component c, Graphics g, int x, int y)
        {
            BASE_ICON.paintIcon(c, g, x, y);
            if (state == State.INDETERMINATE)
            {
                g.setColor(boxColor);
                g.fillRect((int)Math.floor(x + getIconWidth()*(1 - ICON_FILL)/2), (int)Math.floor(y + getIconHeight()*(1 - ICON_FILL)/2),
                           (int)Math.ceil(getIconWidth()*ICON_FILL), (int)Math.ceil(getIconHeight()*ICON_FILL));
            }
        }
    }

    /**
     * State of this TristateCheckBox.
     */
    private State state;

    /**
     * Create a new TristateCheckBox with the specified text and state.
     * 
     * @param text text to show next to the check box
     * @param s initial state of the check box
     */
    public TristateCheckBox(String text, State s)
    {
        super(text, s != State.UNSELECTED);
        setState(s);
        setIcon(new IndeterminateIcon(UIManager.getColor("CheckBox.foreground")));
        setRolloverIcon(new IndeterminateIcon(UIManager.getColor("CheckBox.foreground")));
        addActionListener(new ClickListener());
    }

    /**
     * Create a TristateCheckBox with the specified text in unselected state.
     * 
     * @param text text to show next to the check box
     */
    public TristateCheckBox(String text)
    {
        this(text, State.UNSELECTED);
    }

    /**
     * Create a TristateCheckBox with no text in unselected state.
     */
    public TristateCheckBox()
    {
        this("");
    }

    /**
     * @return The state of this TristateCheckBox.
     */
    public State getState()
    {
        return state;
    }

    /**
     * Check whether or not this TristateCheckBox is indeterminate.
     * 
     * @return <code>true</code> if this TristateCheckBox is indeterminate,
     * and <code>false</code> otherwise.
     */
    public boolean isPartial()
    {
        return state == State.INDETERMINATE;
    }

    /**
     * @inheritDoc
     * 
     * Also update the state to reflect whether or not the box is selected.
     * A state of indeterminate cannot be reached this way; to set it, use
     * #{@link #setState(State)}.
     */
    @Override
    public void setSelected(boolean b)
    {
        setState(b ? State.SELECTED : State.UNSELECTED);
    }

    /**
     * Set the state of this TristateCheckBox.  Also updates the selected
     * state.  A state of indeterminate is considered unselected.
     */
    public void setState(State s)
    {
        state = s;
        super.setSelected(state == State.SELECTED);
    }
}

File: src/main/scala/editor/gui/generic/TristateCheckBox.scala
package editor.gui.generic

import java.awt.Color
import java.awt.Component
import java.awt.Graphics
import java.awt.event.ActionListener
import javax.swing.Icon
import javax.swing.JCheckBox
import javax.swing.UIManager

/**
 * Possible states for a [[TristateCheckBox]]: selected (checked), indeterminate or mixed (displays a box), and unselected (unchecked).
 * @author Alec Roelke
 */
enum TristateCheckBoxState { case Selected, Indeterminate, Unselected }

/**
 * A check box that, in addition to being checked or unchecked, can be in a third state that indicates a mixture of being checked or
 * unchecked, or an indeterminate state.  This state cannot be reached by clicking, only by creating one already in that state. If the
 * box is in the state and is clicked, it becomes unselected.
 * 
 * @constructor create a new tristate check box
 * @param text label next to the check box
 * @param s initial state of the check box
 * 
 * @author Alec Roelke
 */
class TristateCheckBox(text: String = "", private var s: TristateCheckBoxState = TristateCheckBoxState.Unselected) extends JCheckBox(text) {
  import TristateCheckBoxState._

  state = s

  private class IndeterminateIcon(color: Color) extends Icon {
    private val BaseIcon = UIManager.getIcon("CheckBox.icon")
    private val IconFill = 2.0/3.0

    override def getIconHeight = BaseIcon.getIconHeight
    override def getIconWidth = BaseIcon.getIconWidth

    override def paintIcon(c: Component, g: Graphics, x: Int, y: Int) = {
      BaseIcon.paintIcon(c, g, x, y)
      if (state == Indeterminate) {
        g.setColor(color)
        g.fillRect(math.floor(x + getIconWidth*(1 - IconFill)/2).toInt, math.floor(y + getIconHeight*(1 - IconFill)/2).toInt,
                   math.ceil(getIconWidth*IconFill).toInt, math.ceil(getIconHeight*IconFill).toInt)
      }
    }
  }

  /** Listeners for changes in the check box's state. */
  val listeners = collection.mutable.Set[ActionListener]()

  setIcon(IndeterminateIcon(UIManager.getColor("CheckBox.foreground")))
  setRolloverIcon(IndeterminateIcon(UIManager.getColor("CheckBox.foreground")))
  super.addActionListener((e) => {
    state = state match {
      case Selected | Indeterminate => Unselected
      case Unselected => Selected
    }
    listeners.foreach(_.actionPerformed(e))
  })

  /** @return the current state of the check box */
  def state = s

  /**
   * Set the state of the check box.  Triggers action listeners.  This is the only way to set the check box to
   * an indeterminate state after it's been constructed.
   *
   * @param s new state of the check box
   */
  def state_=(s: TristateCheckBoxState) = {
    this.s = s
    super.setSelected(state == Selected)
  }

  /** @return true if the check box is in an indeterminate/multi state, or false otherwise */
  def isPartial = state == Indeterminate

  override def isSelected = state == Selected
  override def setSelected(b: Boolean) = state = if (b) Selected else Unselected
  @deprecated override def addActionListener(l: ActionListener) = listeners += l
  @deprecated override def removeActionListener(l: ActionListener) = listeners -= l
  @deprecated override def getActionListeners = listeners.toArray
}